src/test/shell/bazel/bazel_rules_test.sh

```diff
@@ -743,4 +743,153 @@ EOF
+)
+exit /B 101
+EOF
+    cat >pkg/rules.bzl <<'EOF'
+_SCRIPT_EXT = ".bat"
+_SCRIPT_CONTENT = """
+if not defined RUN_UNDER_VALUE (
+  echo run_under_env value is not set
+  exit /B 1
+)
+start /b /wait bash.exe -c "%RUN_UNDER_VALUE%"
+if %ERRORLEVEL% NEQ 101 (
+  echo Failed to execute run_under_env value
+  exit /B 1
+)
+exit /B 0
+"""
+EOF
+  else
+  local run_under_script="pkg/run_under.sh"
+  echo 'exports_files(["run_under.sh"])' >> pkg/BUILD
+    cat >pkg/run_under.sh <<'EOF'
+#!/bin/bash
+if [[ -z "${RUN_UNDER_VALUE}" ]]; then
+  echo --run_under script was unexpectedly executed
+fi
+exit 101
+EOF
+    chmod +x pkg/run_under.sh
+    cat >pkg/rules.bzl <<'EOF'
+_SCRIPT_EXT = ".sh"
+_SCRIPT_CONTENT = """#!/bin/bash
+if [[ -z "${RUN_UNDER_VALUE:-}" ]]; then
+  echo "run_under_env value is not set"
+  exit 1
+fi
+${RUN_UNDER_VALUE}
+if [[ $? != 101 ]]; then
+  echo "Failed to execute run_under_env value"
+  exit 1
+fi
+exit 0
+"""
+EOF
+  fi
+
+  cat >>pkg/rules.bzl <<'EOF'
+def _my_executable_impl(ctx):
+    executable_sh = ctx.actions.declare_file(ctx.attr.name + _SCRIPT_EXT)
+    ctx.actions.write(
+        output = executable_sh,
+        content = _SCRIPT_CONTENT,
+        is_executable = True,
+    )
+    return [
+        DefaultInfo(
+            executable = executable_sh,
+        ),
+        RunEnvironmentInfo(
+            run_under_env = "RUN_UNDER_VALUE",
+        ),
+    ]
+
+my_executable = rule(
+    implementation = _my_executable_impl,
+    attrs = {},
+    executable = True,
+)
+
+my_test = rule(
+    implementation = _my_executable_impl,
+    attrs = {},
+    test = True,
+)
+
+def _my_wrapper_impl(ctx):
+    executable_sh = ctx.actions.declare_file(ctx.attr.name + _SCRIPT_EXT)
+    ctx.actions.expand_template(
+        output = executable_sh,
+        template = ctx.file._script,
+        substitutions = {},
+        is_executable = True,
+    )
+    return [DefaultInfo(
+        executable = executable_sh,
+    )]
+
+my_wrapper = rule(
+    implementation = _my_wrapper_impl,
+    attrs = {
+        "_script": attr.label(
+            allow_single_file = True,
+            default = Label("//pkg:run_under" + _SCRIPT_EXT),
+        ),
+    },
+    executable = True,
+)
+EOF
+
+  cat >.bazelrc <<'EOF'
+build --announce_rc
+build --subcommands
+test --test_output=errors
+# https://docs.bazel.build/versions/main/windows.html
+startup --windows_enable_symlinks
+build --enable_runfiles
+EOF
+
+  # Test --run_under as a custom command
+  bazel test //pkg:my_test --run_under="$(pwd)/${run_under_script}" >$TEST_log 2>&1 \
+    || fail "Test binary with command should have exit code 0"
+
+  # Test --run_under as a target
+  bazel test //pkg:my_test --run_under=//pkg:my_wrapper >$TEST_log 2>&1 \
+    || fail "Test binary with target should have exit code 0"
+
+  # Test --run_under as a custom command
+  bazel run //pkg:my_executable --run_under="$(pwd)/${run_under_script}" >$TEST_log 2>&1 \
```

Comment:
This test looks pretty thorough!

Only thing I see missing is test coverage for `--run_under` options (`--run_under=//pkg:debugger --attach`) with `run_under_env`.

Reply:
Good catch, updated!